src/sage/rings/valuation/gauss_valuation.py

```diff
@@ -343,7 +343,9 @@ def reduce(self, f, check=True, degree_bound=None, coefficients=None, valuations
             f = f.truncate(degree_bound + 1)
 
         try:
-            return f.map_coefficients(self._base_valuation.reduce, self._base_valuation.residue_field())
+            if f.map_coefficients(self._base_valuation.reduce, self._base_valuation.residue_field())!= 0:
+                return f.map_coefficients(self._base_valuation.reduce, self._base_valuation.residue_field())
+            raise ValueError("Cannot reduce element with finite valuation; use reduce_to_unit instead")
         except Exception:
             if check and not all(v >= 0 for v in self.valuations(f)):
                 raise ValueError("reduction not defined for non-integral elements and %r is not integral over %r" % (f, self))
```

Comment:
```suggestion
            reduction = f.map_coefficients(self._base_valuation.reduce, self._base_valuation.residue_field())
        except Exception:
            if check and not all(v >= 0 for v in self.valuations(f)):
                raise ValueError("reduction not defined for non-integral elements and %r is not integral over %r" % (f, self))
            raise
        if not reduction:
            raise ValueError("cannot reduce element with finite valuation; use reduce_to_unit instead")
        return reduction
```
This should be tested, too.  (I did not look at all into what you are actually doing, I'm just a syntax bot)